src/dmd/typesem.d

```diff
@@ -1650,13 +1650,42 @@ private extern (C++) final class TypeSemanticVisitor : Visitor
         import dmd.traits : semanticTraits;
 
         result = null;
+        if (mtype.ty == Terror)
+        {
+            result = mtype;
+            return;
+        }
+        if (mtype.exp.ident != Id.getMember
+            && mtype.exp.ident != Id.parent
+            && mtype.exp.ident != Id.getOverloads
+            && mtype.exp.ident != Id.getVirtualFunctions
+            && mtype.exp.ident != Id.getVirtualMethods
+            && mtype.exp.ident != Id.getAttributes)
+        {
+            static immutable (const(char)*)[2] ctxt = ["as type", "in alias"];
+            .error(mtype.loc, "trait `%s` is either invalid or not supported %s",
+                 mtype.exp.ident.toChars, ctxt[mtype.inAliasDeclaration]);
+            result = mtype;
+            mtype.ty = Terror;
+            return;
+        }
         if (Expression e = semanticTraits(mtype.exp, sc))
         {
-            if (Dsymbol ds = getDsymbol(e))
+            if (TupleExp te = e.toTupleExp)
+                mtype.sym = new TupleDeclaration(mtype.loc,
+                    Identifier.generateIdWithLoc("__tup", mtype.loc), cast(Objects*) te.exps);
```

Comment:
profiling has show that `generateIdWithLoc` is the culprit for a serious slowdown. I don't know if it's known but this function should not be used. Apparently it would try to expand a filename and/or would lookup in the pool too much.